src/core/ddsi/src/q_plist.c

```diff
@@ -2895,44 +2911,42 @@ static void fill_property(dds_property_t *prop, const char *name, const char *va
   prop->propagate = false;
 }
 
-void nn_xqos_mergein_security_config (dds_qos_t *xqos, const struct omg_security_configuration_type *cfg)
+/**
+ * Add DDS Security configuration to the QoS as a Property policy used by the security
+ * plugins to get their proper settings. If security properties are already present in
+ * the QoS, the settings from configuration are ignored.
+ */
+bool nn_xqos_mergein_security_config (dds_qos_t *xqos, const struct omg_security_configuration_type *cfg)
 {
-  if (cfg)
-  {
-    /* Add DDS Security configuration to the QoS as a Property policy,
-     * used by the security plugins to get their proper settings. */
-    dds_property_t *properties;
-    uint32_t idx = 0;
-
-    /* No property should be present yet. */
-    assert(!(xqos->present & QP_PROPERTY_LIST));
+  assert(cfg != NULL);
 
-    /* Prepare properties. */
-    properties = ddsrt_malloc(8 /* max */ * sizeof (dds_property_t));
+  if (!(xqos->present & QP_PROPERTY_LIST))
+  {
+    xqos->property.value.n = 0;
+    xqos->property.value.props = NULL;
+    xqos->property.binary_value.n = 0;
+    xqos->property.binary_value.props = NULL;
+    xqos->present |= QP_PROPERTY_LIST;
+  }
 
-    /* Fill properties. */
-    fill_property(&(properties[idx++]), "dds.sec.auth.identity_ca", cfg->authentication_properties.identity_ca);
-    fill_property(&(properties[idx++]), "dds.sec.auth.private_key", cfg->authentication_properties.private_key);
-    fill_property(&(properties[idx++]), "dds.sec.auth.identity_certificate", cfg->authentication_properties.identity_certificate);
-    fill_property(&(properties[idx++]), "dds.sec.access.permissions_ca", cfg->access_control_properties.permissions_ca);
-    fill_property(&(properties[idx++]), "dds.sec.access.governance", cfg->access_control_properties.governance);
-    fill_property(&(properties[idx++]), "dds.sec.access.permissions", cfg->access_control_properties.permissions);
+  /* check for existing security properties (name starts with dds.sec. conform DDS Security spec 7.2.4.1) */
+  if (!nn_xqos_has_prop (xqos, "dds.sec.", true))
```

Comment:
Stylistically, I much prefer conditionals where one branch is trivial and the other quite involved such that the short branch comes first. Like the early-outs, it reduces the mental state. But like the other comment, this one too is left to your discretion 🙂

Reply:
Fixed